packages/paymaster/src/BiconomyPaymaster.ts

```diff
@@ -1,49 +1,246 @@
+        Logger.log('passed preferred token is ', paymasterServiceData.tokenInfo.preferredToken)
+        preferredToken = paymasterServiceData.tokenInfo.preferredToken
+      }
+    }
+
+    Logger.log('userop is ', userOp)
+
+    if (
+      paymasterServiceData.tokenInfo &&
+      paymasterServiceData.tokenInfo.tokenList &&
+      paymasterServiceData.tokenInfo.tokenList.length != 0
+    ) {
+      feeTokensArray = paymasterServiceData.tokenInfo.tokenList
+    }
+
+    const feeQuotes: Array<PaymasterFeeQuote> = []
+
+    if (paymasterServiceData.sponsorshipInfo) {
+      if (paymasterServiceData.sponsorshipInfo.webhookData) {
+        webhookData = paymasterServiceData.sponsorshipInfo.webhookData
+      }
+
+      // Could check here that this must be provided
+      if (paymasterServiceData.sponsorshipInfo.smartAccountInfo) {
+        smartAccountInfo = paymasterServiceData.sponsorshipInfo.smartAccountInfo
+      }
+    }
+
+    try {
+      const response: PaymasterServiceSuccessResponse = await sendRequest({
+        url: `${this.paymasterConfig.paymasterUrl}`,
+        method: HttpMethod.Post,
+        body: {
+          method: 'pm_getFeeQuoteOrData',
+          params: [
+            userOp,
+            {
+              ...(mode !== null && { mode }),
+              tokenInfo: {
+                tokenList: feeTokensArray,
+                ...(preferredToken !== null && { preferredToken })
+              },
+              sponsorshipInfo: {
+                ...(webhookData !== null && { webhookData }),
+                smartAccountInfo: smartAccountInfo
+              }
+            }
+          ], // As per current API
+          id: getTimestampInSeconds(),
+          jsonrpc: '2.0'
+        }
+      })
+
+      if (response && response.result) {
+        Logger.log('feeInfo ', response.result)
+        if (response.result.mode == PaymasterMode.ERC20) {
+          const feeQuotesResponse: Array<PaymasterFeeQuote> = response.result.feeQuotes
+          const paymasterAddress: string = response.result.paymasterAddress
+          // check all objects iterate and populate below calculation for all tokens
+          return { feeQuotes: feeQuotesResponse, tokenPaymasterAddress: paymasterAddress }
+        } else if (response.result.mode == PaymasterMode.SPONSORED) {
+          const paymasterAndData: string = response.result.paymasterAndData
+          return paymasterAndData
+        } else {
+          throw new Error('Failed to fetch feeQuote or paymaster data')
+        }
+      } else {
+        // Review if any case response.ok could be true and response would be below
+        /*{
+          "statusCode": 500,
+          "message": "Internal server error"
+        }*/
+        // Note: we may not throw if we include strictMode off and return paymasterData '0x'.
+        // pm service could handle this case by case. Needs Review
+        throw new Error('Failed to fetch feeQuote or paymaster data')
```

Comment:
throw the response.error object here.

Reply:
here in utils we're already throwing if response has error property

```
if (response.ok) {
    if (jsonResponse && jsonResponse.hasOwnProperty('result')) {
      return jsonResponse as T
    } else if (jsonResponse && jsonResponse.hasOwnProperty('error')) {
      const error = jsonResponse.error
      throw new Error(`JSON-RPC error: ${error}`)
    } else {
      throw new Error('Invalid JSON-RPC response')
    }
  }
```

so in our file it will reach our catch block where it is rethrown

```
} catch (error: any) {
      Logger.error("can't query fee quotes - reason: ", error)
      throw new Error('Failed to fetch feeQuote or paymaster data' + error.toString())
    }
```